sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala

```diff
@@ -41,120 +42,138 @@ object ScalaNativePluginInternal {
+      platformDepsCrossVersion := ScalaNativeCrossVersion.binary,
+      nativeClang := nativeConfig.value.clang.toFile,
+      nativeClangPP := nativeConfig.value.clangPP.toFile,
+      nativeCompileOptions := nativeConfig.value.compileOptions,
+      nativeLinkingOptions := nativeConfig.value.linkingOptions,
+      nativeMode := nativeConfig.value.mode.name,
+      nativeGC := nativeConfig.value.gc.name,
+      nativeLTO := nativeConfig.value.LTO.name,
+      nativeLinkStubs := nativeConfig.value.linkStubs,
+      nativeCheck := nativeConfig.value.check,
+      nativeDump := nativeConfig.value.dump
+    )
+  }
 
-  lazy val scalaNativeGlobalSettings: Seq[Setting[_]] = Seq(
-    nativeWarnOldJVM := {
-      val logger = streams.value.log
-      Try(Class.forName("java.util.function.Function")).toOption match {
-        case None =>
-          logger.warn("Scala Native is only supported on Java 8 or newer.")
-        case Some(_) =>
-          ()
-      }
-    },
-    onComplete := {
-      val prev: () => Unit = onComplete.value
-      () => {
-        prev()
-        testAdapters.getAndSet(Nil).foreach(_.close())
+  lazy val scalaNativeGlobalSettings: Seq[Setting[_]] = {
+    val config = build.NativeConfig.empty
+      .withClang(interceptBuildException(Discover.clang()))
+      .withClangPP(interceptBuildException(Discover.clangpp()))
+      .withCompileOptions(Discover.compileOptions())
+      .withLinkingOptions(Discover.linkingOptions())
+      .withLTO(Discover.LTO())
+      .withGC(Discover.GC())
+      .withMode(Discover.mode())
+      .withOptimize(Discover.optimize())
+
+    Seq(
+      nativeConfig := config,
+      nativeBuildConfig := build.Config.empty
+        .withCompilerConfig(config),
+      nativeWarnOldJVM := {
+        val logger = streams.value.log
+        Try(Class.forName("java.util.function.Function")).toOption match {
+          case None =>
+            logger.warn("Scala Native is only supported on Java 8 or newer.")
+          case Some(_) =>
+            ()
+        }
+      },
+      onComplete := {
+        val prev: () => Unit = onComplete.value
+        () => {
+          prev()
+          testAdapters.getAndSet(Nil).foreach(_.close())
+        }
       }
-    }
-  )
+    )
+  }
 
-  lazy val scalaNativeConfigSettings: Seq[Setting[_]] = Seq(
-    nativeTarget := interceptBuildException {
-      val cwd   = nativeWorkdir.value.toPath
-      val clang = nativeClang.value.toPath
-      Discover.targetTriple(clang, cwd)
-    },
-    artifactPath in nativeLink := {
-      crossTarget.value / (moduleName.value + "-out")
-    },
-    nativeWorkdir := {
-      val workdir = crossTarget.value / "native"
-      if (!workdir.exists) {
-        IO.createDirectory(workdir)
-      }
-      workdir
-    },
-    nativeConfig := {
-      val mainClass = selectMainClass.value.getOrElse {
-        throw new MessageOnlyException("No main class detected.")
+  def scalaNativeConfigSettings(key: TaskKey[File]): Seq[Setting[_]] = {
```

Comment:
This is only called with `key = nativeLink`, so you can just use `nativeLink` instead. In Scala.js we have two tasks, `fastOptJS` and `fullOptJS`, so it makes sense. But Scala Native only has the one `nativeLink`. That also means that most `in key`s are actually useless can be removed instead of writing `in nativeLink`.